pfcpiface/bess.go

```diff
@@ -1010,6 +1030,148 @@ func (b *bess) delCounter(ctx context.Context, done chan<- bool, ctrID uint32, c
+			cir = 1                         // Mark all traffic as yellow
+			pir = meterConfig.N6RateBps / 8 // bit/s to byte/s
+		} else {
+			gate = sliceMeterGateUnmeter
+		}
+		if meterConfig.N6BurstBytes != 0 {
+			cbs = 1 // Mark all traffic as yellow
+			pbs = meterConfig.N6BurstBytes
+			ebs = 0 // Unused
+		} else {
+			cbs = DefaultBurstSize
+			pbs = DefaultBurstSize
+			ebs = 0 // Unused
+		}
+		q := &pb.QosCommandAddArg{
+			Gate:      gate,
+			Cir:       cir, /* committed info rate */
+			Pir:       pir, /* peak info rate */
+			Cbs:       cbs, /* committed burst size */
+			Pbs:       pbs, /* Peak burst size */
+			Ebs:       ebs, /* Excess burst size */
+			DeductLen: 0,   /* Exclude Ethernet header */
+			Fields: []*pb.FieldData{
+				intEnc(uint64(farForwardU)), /* Action */
+				intEnc(uint64(0)),           /* tunnel_out_type */
+			},
+		}
+		any, err = anypb.New(q)
+		if err != nil {
+			log.Errorln("Error marshalling the rule", q, err)
+			return
+		}
+		b.processSliceMeter(ctx, any, upfMsgTypeAdd)
+
+		// Uplink N9 slice meter config
+		if meterConfig.N9RateBps != 0 {
+			gate = sliceMeterGateMeter
+			cir = 1                         // Mark all traffic as yellow
+			pir = meterConfig.N9RateBps / 8 // bit/s to byte/s
+		} else {
+			gate = sliceMeterGateUnmeter
+		}
+		if meterConfig.N9BurstBytes != 0 {
+			cbs = 1 // Mark all traffic as yellow
+			pbs = meterConfig.N9BurstBytes
+			ebs = 0 // Unused
+		} else {
+			cbs = DefaultBurstSize
+			pbs = DefaultBurstSize
+			ebs = 0 // Unused
+		}
+		q = &pb.QosCommandAddArg{
+			Gate:      gate,
+			Cir:       cir, /* committed info rate */
+			Pir:       pir, /* peak info rate */
+			Cbs:       cbs, /* committed burst size */
+			Pbs:       pbs, /* Peak burst size */
+			Ebs:       ebs, /* Excess burst size */
+			DeductLen: 0,   /* Exclude Ethernet header */
+			Fields: []*pb.FieldData{
+				intEnc(uint64(farForwardU)), /* Action */
+				intEnc(uint64(1)),           /* tunnel_out_type */
+			},
+		}
+		any, err = anypb.New(q)
+		if err != nil {
+			log.Errorln("Error marshalling the rule", q, err)
+			return
+		}
+		b.processSliceMeter(ctx, any, upfMsgTypeAdd)
+
+		// Downlink N3 slice meter config
+		if meterConfig.N3RateBps != 0 {
+			gate = sliceMeterGateMeter
+			cir = 1                         // Mark all traffic as yellow
+			pir = meterConfig.N3RateBps / 8 // bit/s to byte/s
+		} else {
+			gate = sliceMeterGateUnmeter
+		}
+		if meterConfig.N3BurstBytes != 0 {
+			cbs = 1 // Mark all traffic as yellow
+			pbs = meterConfig.N3BurstBytes
+			ebs = 0 // Unused
+		} else {
+			cbs = DefaultBurstSize
+			pbs = DefaultBurstSize
+			ebs = 0 // Unused
+		}
+		q = &pb.QosCommandAddArg{
+			Gate:      gate,
+			Cir:       cir, /* committed info rate */
+			Pir:       pir, /* peak info rate */
+			Cbs:       cbs, /* committed burst size */
+			Pbs:       pbs, /* Peak burst size */
+			Ebs:       ebs, /* Excess burst size */
+			DeductLen: 50,  /* Exclude Ethernet,IP,UDP,GTP header */
```

Comment:
We do support GTPU extension header(s) too. What will happen then?

Reply:
Good point. I see two options:
1. Passing `deduct_len` per packet, so we can use different values depending on the presence of GTPU extension header. Not sure if possible with the existing QoS module, might require some changes (e.g., to read `deduct_len` from packet metadata)
2. Read `deduct_len` from the JSON config. This would work only if all traffic seen by the same BESS-UPF instance will have GTP-U extensions or not. This is a fair assumption for now in Aether, where we use different UPFs for 5G and 4G base stations, but I'm not sure if this is a valid assumption for Intel's use cases.

Whatever options we decide to pursue, I suggest we push this to the backlog (the SD-Fabric team can take care of this), leave 50 (since we mostly work with 4G base stations for now) and put a FIXME comment.

Reply:
Added a `todo` right above the entry.

Reply:
Tracked here: https://jira.opennetworking.org/browse/SDFAB-571

Reply:
We can solve this simply by making sure the config talks about the limits from the wire perspective as described here right - 
https://github.com/omec-project/upf-epc/pull/324#issuecomment-921454949

Reply:
Accordingly we can use deduct_len be ` -24 Bytes`. This way we do not have to bother if there way extension header or not, encapsulation or not.